python/cog/predictor.py

```diff
@@ -81,51 +71,69 @@ def predict(self, **kwargs: Any) -> Any:
 
 
 def run_setup(predictor: BasePredictor) -> None:
+    weights = get_weights_argument(predictor)
+    if weights:
+        predictor.setup(weights=weights)
+    else:
+        predictor.setup()
+
+
+async def run_setup_async(predictor: BasePredictor) -> None:
+    weights = get_weights_argument(predictor)
+    maybe_coro = predictor.setup(weights=weights) if weights else predictor.setup()
+    if maybe_coro:
+        return await maybe_coro
+
+
+def get_weights_argument(
+    predictor: BasePredictor,
+) -> Union[CogFile, CogPath, str, None]:
+    # by the time we get here we assume predictor has a setup method
     weights_type = get_weights_type(predictor.setup)
 
     # No weights need to be passed, so just run setup() without any arguments.
     if weights_type is None:
         predictor.setup()
         return
 
-    weights: Union[io.IOBase, Path, str, None]
-
     weights_url = os.environ.get("COG_WEIGHTS")
-    weights_path = "weights"
+    weights_path = "weights"  # this is the source of a bug isn't it?
```

Comment:
?

Reply:
https://www.notion.so/replicate/Cog-Setup-Path-Problem-2fc41d40bcaf47579ccd8b2f4c71ee24

Reply:
besides downloading the weights twice as mentioned in that notion doc, there's a lot of code that refers to a bug that I don't quite understand

https://github.com/replicate/cog-flan-models/blob/f35f162f03b2559a755c5ffbe47a5f17009ab867/predict.py#L168-L170

https://github.com/replicate/cog-mpt-7b-storywriter-65k/blob/main/predict.py#L41-L43

https://github.com/replicate/cog-llama/blob/main/predict.py#L18-L20

https://github.com/replicate/cog-sdxl/blob/main/predict.py#L184-L185

https://github.com/replicate/cog-llama-template/blob/main/predict.py#L44-L46